Unify supported models and static predictions (#1835)
deepchecks/deepchecks: docs/source/user-guide/tabular/supported_models.rst

.. _supported_models:

**************************************
Supported Models and Static Prediction
**************************************

Some checks, including all the model evaluation related checks, require model predictions in order to run.
The prediction can be provided **either** within a model object or as pre-computed static predictions.

> Whether provided from a model interface or as a pre-computed static predictions,

Reviewer: Add a title to this short section to make it clear + stand out

the predictions must be in the following format based on the task type:

* Classification: ``proba``, an `ndarray` of shape ``(n_samples, n_classes)`` containing the predicted
  probability of each class per sample in the dataset.
* Regression: ``pred``, an `ndarray` of shape ``(n_samples,)``, containing the predicted value
  for each sample in the dataset.

For classification tasks, it is optional to also supply ``pred`` predictions. If not supplied, Deepchecks will
assume the predicted value is the class with the highest probability.

> Using a Model Interface

Reviewer: My main comment here (many of the others are small...)
is that I don't feel it's clear for a user "which models are supported?"
and "how can I make my model supported if it isn't already?" (e.g. how to copy paste simple code for model wrapper, wrap their model given their task type, and return that object...)

***********************

Supported model objects should implement at least one of the following interfaces based on the task type:

* Classification: ``predict_proba`` method which receives an |array-like|  of shape ``(n_samples, n_features)``
  and is expected to return an `ndarray` as specified above.
* Regression: ``predict`` method which receives an |array-like|  of shape ``(n_samples, n_features)``
  and is expected to return an `ndarray` as specified above.

.. Note::
    The methods are implemented using the scikit-learn api conventions therefore built-in scikit-learn
    classifiers and regressors,
    along with many additional popular models types (e.g. XGBoost, LightGBM, CatBoost etc.) are supported.

> Optional Model Interface

Reviewer: These are less interesting than the code examples for "wrappers", would put them in a less major place

>>> class simple_regression_with_importance_model:
...     def predict(X: pd.DataFrame) -> pd.Series:
...         ...
...     @property
...     def feature_importances_(self):
...         ...

Running Deepchecks with a Supported Model
=========================================

>>> from deepchecks.tabular.datasets.classification.iris import load_data, load_fitted_model
... from deepchecks.tabular.suites import model_evaluation
...
... ds_train, ds_test = load_data(data_format='Dataset')
... rf_clf = load_fitted_model() # trained sklearn RandomForestClassifier
... model_evaluation().run(train_dataset=ds_train, test_dataset=ds_test, model=rf_clf)

Using Pre-computed Predictions
******************************

There are several cases in which it's useful to provide pre-computed predictions instead of a model object.

For example, in cases where the model's prediction process takes a long time or is computationally expensive.
Another example is when the model is stored in different location or does not have a compatible API for providing
the predictions.

.. Note::
    If the train dataset shares indices with the test dataset we will add train/test prefixes to the indexes.

> Running Deepchecks with Pre-computed Predictions

Reviewer: I think this section is missing some "motivation"
(about the option to sending prediction instead of model. Alternatively (maybe better) this can be explained at the top.

Author: Updated on top, see what you think

================================================

> We will run the Deepchecks model evaluation suite using static predictions from a random forest classification model.

Reviewer: I don't think the term "static predictions" is clear

> Since it is a classification model we will supply the ``proba``  probabilities predictions as defined above.

Reviewer: same here about ``proba``

Reviewer: "as defined above" - they probably didn't read, since it's under a "Using Model Interface" section.
Possibly - direct link + separate the "required api" and the "supported model / wrapper" vs. "model predictions" options?

> In addition, we will calculate and pass |permutation importance|.

Reviewer: Short motivation? (about why to pass this, is it necessary, etc.)

>>> from deepchecks.tabular.datasets.classification.iris import load_data, load_fitted_model
... from deepchecks.tabular.suites import model_evaluation

> ... from deepchecks.tabular.feature_importance import calculate_feature_importance

Reviewer: I think it's something we really want to ecnourage someone sending us predictions to send, so I would still add it here (with some motivation explanation), even if it is somewhere else

.. _supported_models:

===================================
Working with Models and Predictions
===================================

> Some checks, specially the model evaluation related checks, require model predictions in order to run.

Reviewer: ```suggestion
Some checks, mainly the ones related to model evaluation, require model predictions in order to run.
```

* **Probabilities per class**: should be provided as an |array-like| of shape ``(n_samples, n_classes)``
  containing the predicted probability of each class per sample. Probabilities per class are only required for
  classification tasks.

.. _supported_models__passing_a_model:

Passing a Model
===============

Deepchecks requires models to follow the |scikit-learn API conventions| for calculating predicted values
and probabilities per class. Therefore built-in scikit-learn classifiers and regressors,
along with many additional popular models types (e.g. XGBoost, LightGBM, CatBoost etc.) are supported out of the box.

Specifically, deepchecks requires the following methods to be implemented in the model object:

* ``predict`` method which receives an |array-like|  of shape ``(n_samples, n_features)`` containing the
  input features and returns :ref:`predicted values <supported_models__predictions_format>`.
* ``predict_proba`` method which receives an |array-like|  of shape ``(n_samples, n_features)`` containing the
  input features and returns :ref:`probabilities per class <supported_models__predictions_format>`.
  This method is required only for classification tasks.

> Running Deepchecks with a Supported Model

Reviewer: ```suggestion
Running Deepchecks With a Supported Model
```

-----------------------------------------

.. literalinclude:: ../../../../examples/examples_supported_models.py
    :language: python
    :lines: 0-7
    :tab-width: 0

> Adapting Your Model
> -------------------

Reviewer: ```suggestion
Adapting Your Custom Model
------------------------------
```

>>> class MyModelWrapper:
...     def predict(X: pd.DataFrame) -> np.ndarray:
...         # Implement based on base model's API
...         ...
...     def predict_proba(X: pd.DataFrame) -> np.ndarray:
...         # Implement based on base model's API, only required for classification tasks.
...         ...
...     @property
...     def feature_importances_(self) -> pd.Series:  # optional
...         # Return a pandas Series with feature names as index and their corresponding importance as values.
...         ...

Feature Importance (Optional)
-----------------------------

Some checks require the model's
:doc:`feature importance </api/generated/deepchecks.user-guide.tabular.feature_importance>`
for their analysis. By default, if available, it is extracted directly from the model via property
(``feature_importances_`` or ``coef_`` for a linear model) otherwise it is calculated
using |permutation importance|. The required format for the feature importance is a pandas series with feature names
as index and their corresponding importance as values.

.. _supported_models_using_pre-computed_predictions:

Using Pre-computed Predictions
==============================

> The predictions should be passed via the y_proba and y_pred arguments of the suite / check run method in

Reviewer: ```suggestion
The predictions should be passed via the ``y_proba`` and ``y_pred`` arguments of the suite / check's ``run`` method in
```

and y_proba receives the probabilities per class, which is only required for classification tasks.

The predictions should be provided for each dataset supplied to the suite / check. For example the
:doc:`Simple Model Comparison </api/generated/deepchecks.tabular.checks.model_evaluation.SimpleModelComparison>`
check for a regression model
requires both train and test :ref:`predicted values <supported_models__predictions_format>`
to be provided via the y_pred_train, y_pred_test arguments.

For classification tasks, predicted values are not mandatory. If not supplied,
deepchecks will assume the predicted class is the class with the highest predicted probability.

.. Note::
    When using pre-computed predictions, if the train dataset shares indices with the test dataset we
    will add train/test prefixes to the indexes.

Running Deepchecks with Pre-computed Predictions
------------------------------------------------

We will run the deepchecks model evaluation suite using pre-computed predictions from a random forest classification
model. In addition, we will calculate and pass |permutation importance| which provides a better estimate of the

> effect of different features on the model's performance. see
> :doc:`Feature Importance </api/generated/deepchecks.user-guide.tabular.feature_importance>` for more details.

Reviewer: ```suggestion
effect of different features on the model's performance. See the
:doc:`feature importance API reference </api/generated/deepchecks.user-guide.tabular.feature_importance>` for more details.
```
